crates/host/src/cache.rs

```diff
@@ -0,0 +1,588 @@
+            if self.purge_cycle_count.load(Ordering::Relaxed) >= purge_cycles {
+                self.poll_purge_once().await;
+            }
+        }
+    }
+}
+
+impl<K, V> std::fmt::Debug for Cache<K, V>
+where
+    K: std::hash::Hash + Eq + Clone + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Cache")
+            .field("purge_cycles", &self.purge_cycles)
+            .field(
+                "purge_cycle_count",
+                &self.purge_cycle_count.load(Ordering::Relaxed),
+            )
+            .finish()
+    }
+}
+
+impl<K, V> Clone for Cache<K, V>
+where
+    K: std::hash::Hash + Eq + Clone + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    fn clone(&self) -> Self {
+        Self {
+            internal: Arc::clone(&self.internal),
+            purge_cycles: self.purge_cycles,
+            purge_cycle_count: AtomicU32::new(self.purge_cycle_count.load(Ordering::Relaxed)),
+        }
+    }
+}
+
+/// A builder for creating a new cache with configurable settings
+///
+/// Example:
+/// ```
+/// use std::time::Duration;
+/// use wasmcloud_host::cache::Cache;
+///
+/// let cache = Cache::<String, i32>::builder()
+///     .with_ttl(Duration::from_secs(10))
+///     .with_purge_cycles(10)
+///     .build();
+/// ```
+pub struct CacheBuilder<K, V>
+where
+    K: std::hash::Hash + Eq + Clone + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    ttl: Duration,
+    purge_cycles: Option<u32>,
+    lazy: bool,
+    _phantom: PhantomData<(K, V)>,
+}
+
+impl<K, V> Default for CacheBuilder<K, V>
+where
+    K: std::hash::Hash + Eq + Clone + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    fn default() -> Self {
+        Self {
+            ttl: Duration::from_secs(300), // 5 minutes default
+            purge_cycles: None,
+            lazy: true,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<K, V> CacheBuilder<K, V>
+where
+    K: std::hash::Hash + Eq + Clone + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    /// Create a new cache builder with default settings
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Set the time-to-live (TTL) for cache entries
+    pub fn with_ttl(mut self, ttl: Duration) -> Self {
+        self.ttl = ttl;
+        self
+    }
+
+    /// Set the number of operations after which to run a purge cycle
+    /// If None, purging only happens on explicit timer expiration
+    pub fn with_purge_cycles(mut self, cycles: u32) -> Self {
+        self.purge_cycles = Some(cycles);
+        self
+    }
+
+    /// Disable automatic purge cycles (default behavior)
```

Comment:
For the basic use case here, I think it might be better to just default to having an automatic purge (or not even have it configurable). My main concern is that you always create and insert into the `DelayQueue` even if you're not setting up auto purging. I could be misunderstanding why you chose this as the default though.

Any thoughts?

Reply:
I chose that as a default behavior because I was adding extra functionality from our previous discussion, though it was not necessarily needed. But I agree with you. Now, about keeping it configurable, it might be better given that the ideal number of cycles may vary by use case.

Reply:
What it's a good purge cycledefaultl val,I'llill put 1000 ,for now but lmk